src/infra/core/src/compact_service_impl.rs

```diff
@@ -0,0 +1,389 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use chrono::Utc;
+use datafusion::prelude::*;
+use dill::{component, interface};
+use domain::compact_service::{
+    CompactError,
+    CompactionMultiListener,
+    CompactionPhase,
+    InvalidDatasetKindError,
+    NullCompactionListener,
+};
+use futures::stream::TryStreamExt;
+use kamu_core::compact_service::CompactService;
+use kamu_core::*;
+use opendatafabric::{
+    AddData,
+    DatasetHandle,
+    DatasetKind,
+    DatasetName,
+    MetadataEvent,
+    Multihash,
+    OffsetInterval,
+};
+use url::Url;
+
+use crate::utils::random_names_helper::get_random_operation_name_with_prefix;
+use crate::*;
+
+pub struct CompactServiceImpl {
+    dataset_repo: Arc<dyn DatasetRepository>,
+    dataset_authorizer: Arc<dyn domain::auth::DatasetActionAuthorizer>,
+    run_info_dir: PathBuf,
+}
+
+#[derive(Debug, Default, Clone)]
+struct DataSliceBatchInfo {
+    pub data_slices_batch: Vec<Url>,
+    pub prev_offset: Option<u64>,
+    pub new_offset_interval: Option<OffsetInterval>,
+    pub new_file_path: Option<PathBuf>,
+    // Hash of block will not be None value in case
+    // when data_slices_batch.len() == 1
+    // and will be used tp not rewriting such blocks
+    pub hash: Option<Multihash>,
+}
+
+struct ChainFilesInfo {
+    _block_file_urls: Vec<Url>,
+    old_head: Multihash,
+    data_slice_batches: Vec<DataSliceBatchInfo>,
+}
```

Comment:
The algorithm does not consider checkpoints. We should leave the very latest checkpoint only and drop any checkpoints before. If there were no checkpoints, nothing to do.

Reply:
Added checkpoint handling